src/core/qgsmaplayerstore.cpp

```diff
@@ -74,6 +74,11 @@ QList<QgsMapLayer *> QgsMapLayerStore::addMapLayers( const QList<QgsMapLayer *>
       QgsDebugMsg( QStringLiteral( "Cannot add null layers" ) );
       continue;
     }
+    // If the layer is already in the store but its validity has flipped to TRUE reset data source
+    if ( mMapLayers.contains( myLayer->id() ) && ! mMapLayers[myLayer->id()]->isValid() && myLayer->isValid() && myLayer->dataProvider() )
+    {
+      mMapLayers[myLayer->id()]->setDataSource( myLayer->dataProvider()->dataSourceUri(), myLayer->name(), myLayer->providerType(), QgsDataProvider::ProviderOptions() );
```

Comment:
This leaks myLayer - not sure how to correctly hacked this though (it may not be safe to delete the layer here)

Reply:
@nyalldawson does my last commit address your issue?

I could not come up with a better solution.